File: src/features/typing/data/go/minimum_window_substring.go
package main

func minWindow(s string, t string) string {
	need := make(map[byte]int)
	for i := 0; i < len(t); i++ {
		need[t[i]]++
	}
	have := make(map[byte]int)
	required := len(need)
	formed := 0
	left := 0
	best := ""
	bestLen := 1 << 30
	for right := 0; right < len(s); right++ {
		c := s[right]
		have[c]++
		if need[c] > 0 && have[c] == need[c] {
			formed++
		}
		for formed == required {
			length := right - left + 1
			if length < bestLen {
				bestLen = length
				best = s[left : right+1]
			}
			lc := s[left]
			have[lc]--
			if need[lc] > 0 && have[lc] < need[lc] {
				formed--
			}
			left++
		}
	}
	return best
}

File: src/features/typing/data/ts/minimum_window_substring.ts
function minWindow(s: string, t: string): string {
  const need = new Map<string, number>();
  for (const c of t) {
    need.set(c, (need.get(c) ?? 0) + 1);
  }
  const have = new Map<string, number>();
  const required = need.size;
  let formed = 0;
  let left = 0;
  let best = '';
  let bestLen = Infinity;
  for (let right = 0; right < s.length; right++) {
    const c = s[right];
    have.set(c, (have.get(c) ?? 0) + 1);
    if (need.has(c) && have.get(c) === need.get(c)) {
      formed++;
    }
    while (formed === required) {
      const length = right - left + 1;
      if (length < bestLen) {
        bestLen = length;
        best = s.slice(left, right + 1);
      }
      const lc = s[left];
      have.set(lc, have.get(lc)! - 1);
      if (need.has(lc) && have.get(lc)! < need.get(lc)!) {
        formed--;
      }
      left++;
    }
  }
  return best;
}
